### ISSUE TITLE
Why does application.load() assume that ctx.path will be a latin1 string?

### ISSUE THREAD

**Issue by cclauss** [webpy/webpy#687]
[ctx.path = env.get("PATH_INFO").encode("latin1").decode("utf8")](https://github.com/webpy/webpy/blob/master/web/application.py#L456)

**Comment by iredmail** [comment]
i have no idea. Ping @anandology

**Comment by mekarpeles** [comment]
I think there may be a problem with calling web.redirect with the following string:

```
>>> x
'Hélas'
>>> x.encode('latin-1')
b'H\xe9las'
>>> x.encode('latin-1').decode('utf-8')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 1: invalid continuation byte
```

**Comment by mekarpeles** [comment]
Consider this example webpy app which redirects a patron from `/` to `/Hélas`

```
# !/usr/bin/python                                                                                               
# coding=utf-8                                                                                                   

import web

urls = (
    '/', 'Redir',
    '/(.+)', 'Site'
)
app = web.application(urls, globals())

class Redir:
    def GET(self):
        raise web.redirect('/Hélas')

class Site:
    def GET(self, param):
        return param + " works fine"

if __name__ == "__main__":
    app.run()
```

On Python 2 webpy `0.37` it results in 

```
<type 'exceptions.UnicodeDecodeError'> at /
'ascii' codec can't decode byte 0xc3 in position 2: ordinal not in range(128)
```

And on Python 3 webpy `0.61`:

```
  File "/usr/local/lib/python3.5/dist-packages/web/application.py", line 457, in load
    ctx.path = env.get("PATH_INFO").encode("latin1").decode("utf8")
UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 2: invalid continuation byte
```

**Comment by cclauss** [comment]
#516 @schneidersoft Can you please weigh in on [this change](https://github.com/webpy/webpy/commit/146fcbc60610a0aeecefce3598dff9d9706c1a3b#diff-c27618b2fd15107297fecdc40a7fadb2d8b1d4d9d6d02fd3699053a7291242d2R431)?  Why do we want to convert Unicode str into bytes encoded with latin1 and then decode it with utf8?
```
#!/usr/bin/env python3

test_strs = ("AbCdEf", "âêîôû", "Chris Clauß, Life in Lüsai, Graubünden")

for s0 in test_strs:
    print(s0)
    b0 = bytes(s0, "latin-1")
    b1 = s0.encode("latin-1")
    assert b0 == b1, (b0, b1)
    # UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe2 in position 0:
    #     invalid continuation byte
    s1 = b0.decode("utf-8")
    assert s0 == s1, (s0, s1)
    print("Done", s1)
```

**Comment by mekarpeles** [comment]
We're getting around in in Open Library by url quoting whatever unicode is getting passed to the redirect.

**Comment by tfmorris** [comment]
> Why do we want to convert Unicode str into bytes encoded with latin1 and then decode it with utf8?

You don't. And, that will only ever work for pure ASCII. The upper half of ISO Latin-1 (ie ISO 8859-1) doesn't match UTF-8.

You want `urllib.parse.unquote()` instead. Note this fix has already been applied to application.py where the comments say:

>            # Apache and CherryPy webservers unquote the url but lighttpd doesn't.
>            # unquote explicitly for lighttpd to make ctx.path uniform across all servers.

**Comment by pbuckner** [comment]
I can't support the encode().decode(). I think the original suggestion was `bytes(env.get("PATH_INFO"), 'latin1').decode('utf8')` which is slightly different.

Issue I saw was related to using nginx / uwsgi as server. (uwsgi returns latin1 encoded string, as I recall).

Quick test demonstrates you can keep `ctx.path = env.get("PATH_INFO")`, (with no encode/decode)  *and* solve the problem by adding `nginx/` similar to what is done for `lighttpd/`. This allows extended characters to be represented in URL.

`if  env.get("SERVER_SOFTWARE", "").startswith('lighttpd/') or env.get("SERVER_SOFTWARE", "").startswith('nginx/'):`

**Comment by cclauss** [comment]
`str.startswith()` can take a tuple...
`if env.get("SERVER_SOFTWARE", "").startswith(('lighttpd/', 'nginx/'))`

**Comment by cclauss** [comment]
Closed in #690 and released in v0.62
